### PR TITLE
Create named destinations for boxes with -fs-named-destination: create

### LINKED ISSUE #69: Support for named destinations with -fs-named-destination: create;
It'd be nice to resurrect the `-fs-named-destination` support that was originally in flying saucer:

* https://github.com/flyingsaucerproject/flyingsaucer/pull/27
* https://github.com/flyingsaucerproject/flyingsaucer/pull/44

The idea is to create named destinations for elements with an id, and for this to be opt-in with the `-fs-named-destination: create` CSS directive.

The usage in the original PRs is slightly different to my intention, but my use case is the following:

* Create multiple PDFs for separate parts of a document
* Merge together into one final PDF
* Use the named destination to deep link to a named part of the document (e.g. footnotes)

### PR DESCRIPTION
**Pull request by matmannion**
Fixes #69 

Adds support for `-fs-named-destination: create` from Flying Saucer, which was originally implemented in FS with the following two pull requests:

* https://github.com/flyingsaucerproject/flyingsaucer/pull/27
* https://github.com/flyingsaucerproject/flyingsaucer/pull/44

Where there is a box with an `id` attribute, and the `-fs-named-destination: create;` resolved style, it will create a named destination in the document catalog. It doesn't (currently) make any attempt to use these for (e.g.) anchor linking in order to reduce the impact of the change, but this may be useful for implementors who post-process PDFs to be able to get access to a `PDPageXYZDestination` for an id.

### CODE REVIEW

File: openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java
Comment (siegelzc) [inline openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java:44]: I'm a little concerned about performing this linear search every time.

Ideally, I think while the `Box`s are being created, or while the stylesheet is being parsed, we could mark a flag in  `SharedContext`, notifying the `PdfBoxNamedDestinationManager` that this feature is in use. Then, we could ignore the linear search for most PDFs that don't use named destinations.

I looked around for a while and the best option I found was to set the flag inside `CSSParser#declaration`:
```java
                    if (valid) {
+                       if (cssName == CSSName.FS_NAMED_DESTINATION && values.stream().anyMatch(propertyValue -> propertyValue.getIdentValue() == IdentValue.CREATE)) {
+                           ThreadCtx.get().sharedContext().setUsingFsNamedDestination(true);
+                       }
+
                        try {
                            PropertyBuilder builder = CSSName.getPropertyBuilder(cssName);
                            ruleset.addAllProperties(builder.buildDeclarations(
                                    cssName, values, ruleset.getOrigin(), important, !inFontFace));

```

File: openhtmltopdf-examples/src/test/java/com/openhtmltopdf/nonvisualregressiontests/NonVisualRegressionTest.java
Comment (siegelzc) [inline openhtmltopdf-examples/src/test/java/com/openhtmltopdf/nonvisualregressiontests/NonVisualRegressionTest.java:1]: good test

**Review by siegelzc — CHANGES_REQUESTED** [review]

**Comment by siegelzc** [comment]
Once we get this merged, we should update [this wiki page](https://github.com/danfickle/openhtmltopdf/wiki/Custom-CSS-properties)

File: openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java
Comment (matmannion) [reply openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java:64]: This would happen if the PDF being modified has an invalid structure for the names array here, it should be an array of objects that alternate between a string (the name) and an array (representing the destination). If we're in that situation I'm not sure whether we should be overriding what's already there, in case it hides a specific class of bug where we're modifying this earlier in the process, what do you think?

Happy to put the set as suggested if you feel this is more appropriate than leaving as-is.

File: openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java
Comment (matmannion) [reply openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java]: Good shout; done

File: openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java
Comment (matmannion) [reply openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java]: No problem, done.

**Comment by matmannion** [comment]
> Once we get this merged, we should update [this wiki page](https://github.com/danfickle/openhtmltopdf/wiki/Custom-CSS-properties)

We aren't currently hosting a version of this wiki here, but perhaps we should since we're starting to get to the point where it's diverged?

File: openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java
Comment (siegelzc) [reply openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java:64]: Yeah I think your intuition was right. Let's leave it as-is.

**Review by siegelzc — APPROVED** [review]

**Comment by siegelzc** [comment]
> ...tmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java

Good point. I just cloned danfickle's wiki to this repo. Here is the new link: https://github.com/openhtmltopdf/openhtmltopdf/wiki/Custom-CSS-properties

**Comment by siegelzc** [comment]
I'm not sure if GitHub will allow you to click the merge button. If not, LMK when you're ready to send it. Thanks for your contribution.

**Comment by matmannion** [comment]
> I'm not sure if GitHub will allow you to click the merge button. If not, LMK when you're ready to send it. Thanks for your contribution.

I don't have write access (though I'd be interested in becoming a maintainer in general); so ready when you are.

**Pull request merged**
